ZEP000-size-based-snapshot.md

```diff
@@ -0,0 +1,160 @@
+before in atomix in a similar way. Snapshotting can then be triggered quite often until we going under our threshold.
+
+To avoid to take snapshots too often we could have an check in the `AsyncSnapshotDirector` which verifies that the 
+processor position or exporter position has changed with an amount of `Y` positions. We probably need this check to avoid
+to many snapshots at once, since we currently directly replicate these snapshots. This problem mitigates if we build state on follower.
+
+After taking a snapshot we immediately trigger compaction based on the last processed position and exporter position.
+It might happen that the exporter is slow and we are not able to compact yet or not so much. Depending how much we compact we
+will probably triggered again, shortly afterwards. This is good, since we then can overcome our disk issue earlier.
+After the processor or exporter position was for example updated by `Y` positions we will take a snapshot again and
+might be able to compact more then before. If this is not the case then we will be triggered again. But at least we have
+taken a snapshot with higher processed position, which means it will be faster on reprocessing.
+
+The threshold should be configurable and a good value needs probably be evaluated.
+
+**Alternative**
+
+As an alternative we could take snapshot every `X` appends, without taking care of the disk usage, since we expect that we at some point
+are able to compact the log. This would make it probably also much easier, since we are not triggered to often. Here we need to find a good `X`.
+
+Here the question is what happens on no load. I think this is fine, since we don't need to take a snapshot when nothing happens. Eventually
+new appends will arrive and then we will take a snapshot again.
+
+The start up time would benefit of this approach and it would be more predictable how much needs to be reprocessed.
+
+## Compatibility
+
+This should not have any impact of the compatibility.
+
+## Testing
+
+We should test:
+ 
+ * snapshots are taken after a specific disk usage
+ * snapshots are taken again after process or exporter position have changed by `Y`
+ * compaction is done
+
+### Unit Test
+
+Test configuration.
+ 
+### Integration & E2E
+
+Test that snapshots are take after given load is reached and log is compacted afterwards.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Replication could be done more often. Might be an issue on bigger state.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We keep it like it is, but we are not able to handle high load scenarios properly. We will still
+have the issues which are pointed out above.
+
+# Prior art
+[prior-art]: #prior-art
+
+There are several raft implementation out there and I think all of them doing it in the same way, since it 
+is also described in that way in the raft paper see above. I think it was also done in atomix before.
```

Comment:
I think in atomix it takes snapshot periodically (or atleast attempt to take a snapshot). This period was very small - around 10sec. In addition, it also force take a snapshot if it runs out of disk space.